NitroxServer-Subnautica/Serialization/ResourceAssetsParser.cs

```diff
@@ -58,17 +82,457 @@ public static ResourceAssets Parse()
+                                    }
+
+                                    resourcesFile.reader.Position = prevPos; // Reset position
+                                    break;
+                                }
+                            }
+                            break;
+
                     }
                 }
             }
-
             Validate.IsTrue(resourceAssets.WorldEntitiesByClassId.Count > 0);
             Validate.IsTrue(resourceAssets.LootDistributionsJson != "");
 
             return resourceAssets;
         }
 
+        private static AssetFileInfoEx FindComponentOfType<T>(AssetsFileInstance assetsFileInstance, ulong gameObjectPathId)
+        {
+            return FindComponentOfType(typeof(T), assetsFileInstance, gameObjectPathId);
+        }
+
+        private static AssetFileInfoEx FindTransform(AssetsFileInstance assetsFileInstance, ulong gameObjectPathId)
+        {
+            AssetsFileTable assetsTable = assetsFileInstance.table;
+            AssetsFile assetsFile = assetsFileInstance.file;
+            AssetsFileReader afr = assetsFile.reader;
+
+            ulong prevPos = assetsFile.reader.Position;
+            AssetFileInfoEx gameObjectAFI = assetsTable.getAssetInfo(gameObjectPathId);
+            if (gameObjectAFI.curFileType != 1)
+            {
+                Log.Info("An incorrect pathId was passed to FindTransform please check if it points to a proper gameObject: " + gameObjectPathId);
+                return null;
+            }
+
+            afr.Position = gameObjectAFI.absoluteFilePos;
+            afr.Align();
+
+            uint componentCount = afr.ReadUInt32();
+            AssetFileInfoEx componentAFI;
+
+            for (int i = 0; i < componentCount; i++)
+            {
+                uint fileId = (uint)afr.ReadInt32();
+                ulong pathId = (ulong)afr.ReadInt64();
+                if (fileId == 0)
+                {
+                    componentAFI = assetsTable.getAssetInfo(pathId);
+                    if (componentAFI.curFileType == 4) // Ey look we got it
+                    {
+                        afr.Position = prevPos;
+                        return componentAFI;
+                    }
+                }
+                else
+                {
+                    AssetsFileInstance dep = assetsFileInstance.dependencies[(int)fileId - 1];
+                    componentAFI = dep.table.getAssetInfo(pathId);
+                    if (componentAFI.curFileType == 4) // Ey look we got it
+                    {
+                        afr.Position = prevPos;
+                        return componentAFI;
+                    }
+                }
+            }
+
+
+
+            return null; // This should likely never happen
+        }
+
+        private static AssetFileInfoEx FindComponentOfType(Type type, AssetsFileInstance assetsFileInstance, ulong gameObjectPathId)
```

Comment:
Might want these methods in a parsing helper class or the base of an abstract parser